### PR TITLE
Obj loader refactor

### PR DESCRIPTION
**Pull request by AnastaZIuk**

### CODE REVIEW

**Comment by AnastaZIuk** [comment]
> yeah the move is important.
> 
> See the definition of `make_smart_refctd_ptr` its basically a `new` wrapper, you don't want to new here

File: src/irr/asset/CPLYMeshFileLoader.cpp [hunk 1]
```diff
@@ -70,8 +66,7 @@ asset::SAssetBundle CPLYMeshFileLoader::loadAsset(io::IReadFile* _file, const as
         return {};
 
     SContext ctx;
-
-	ctx.File = std::move(std::make_unique<io::IReadFile>(_file));
+	ctx.File = std::move(core::make_smart_refctd_ptr<io::IReadFile>(_file));
```
Comment (devshgraphicsprogramming) [inline src/irr/asset/CPLYMeshFileLoader.cpp]: use raw pointer and don't grab, you won't be using the file anymore after `loadAsset` returns.

File: src/irr/asset/CPLYMeshFileLoader.h [hunk 2]
```diff
@@ -125,18 +128,21 @@ class CPLYMeshFileLoader : public asset::IAssetLoader
 
     struct SContext
     {
-        core::vector<std::unique_ptr<SPLYElement>> ElementList;
-
-		std::unique_ptr<io::IReadFile> File;
-		std::unique_ptr<char> Buffer;
+        core::vector<core::smart_refctd_ptr<SPLYElement>> ElementList;
```
Comment (devshgraphicsprogramming) [inline src/irr/asset/CPLYMeshFileLoader.h]: unique_ptr is the only option here, as SPLYElement does not inherit from IReferenceCounted.

File: src/irr/asset/CPLYMeshFileLoader.h [hunk 3]
```diff
@@ -125,18 +128,21 @@ class CPLYMeshFileLoader : public asset::IAssetLoader
...
+	
+		core::smart_refctd_ptr<io::IReadFile> File;
```
Comment (devshgraphicsprogramming) [inline src/irr/asset/CPLYMeshFileLoader.h]: no need to refcount that.

File: src/irr/asset/CPLYMeshFileLoader.cpp [hunk 4]
```diff
@@ -150,7 +145,7 @@ asset::SAssetBundle CPLYMeshFileLoader::loadAsset(io::IReadFile* _file, const as
 				else
 				{
 					// get element
-					std::unique_ptr<SPLYElement> el = std::move(ctx.ElementList[ctx.ElementList.size()-1]);
+					core::smart_refctd_ptr<SPLYElement> el = std::move(ctx.ElementList[ctx.ElementList.size()-1]);
```
Comment (devshgraphicsprogramming) [inline src/irr/asset/CPLYMeshFileLoader.cpp]: you can use unique_ptr on non-IReferenceCounted objects, in fact `unique_ptr` is great!

File: src/irr/asset/COBJMeshFileLoader.cpp [hunk 5]
```diff
@@ -395,7 +395,9 @@ asset::SAssetBundle COBJMeshFileLoader::loadAsset(io::IReadFile* _file, const as
         auto meshbuffer = core::make_smart_refctd_ptr<asset::ICPUMeshBuffer>();
         mesh->addMeshBuffer(core::smart_refctd_ptr(meshbuffer));
 
-        meshbuffer->getMaterial() = ctx.Materials[m]->Material;
+		// meshbuffer should copy ctx Material
```
Comment (devshgraphicsprogramming) [inline src/irr/asset/COBJMeshFileLoader.cpp:398]: you need to create pair<pipeline,descriptorset> not materials

File: src/irr/asset/COBJMeshFileLoader.cpp [hunk 6]
```diff
@@ -420,9 +422,11 @@ asset::SAssetBundle COBJMeshFileLoader::loadAsset(io::IReadFile* _file, const as
             actualVertexCount = ctx.Materials[m]->Vertices.size();
 
 			{
-				auto indexbuf = core::make_smart_refctd_ptr<asset::ICPUBuffer>(sizeof(uint32_t)*ctx.Materials[m]->Indices.size());
-				memcpy(indexbuf->getPointer(),&ctx.Materials[m]->Indices[0],indexbuf->getSize());
-				desc->setIndexBuffer(std::move(indexbuf));
+				ICPUMeshBuffer::SBufferBinding bufferBinding;
```
Comment (devshgraphicsprogramming) [inline src/irr/asset/COBJMeshFileLoader.cpp:425]: don't fuck around with uninitialized variables, initialize that offset.

File: src/irr/asset/COBJMeshFileLoader.cpp [hunk 7]
```diff
@@ -659,6 +663,7 @@ void COBJMeshFileLoader::readMTL(SContext& _ctx, const char* fileName, const io:
 	mtlReader->read((void*)buf, filesize);
 	const char* bufEnd = buf+filesize;
 
+	// such object should have ICPU/GPURenderpassIndependentPipeline against Material (IGPUMaterial), but how can I get access to material using ICPU/GPURenderpassIndependentPipeline?
```
Comment (devshgraphicsprogramming) [inline src/irr/asset/COBJMeshFileLoader.cpp:666]: There is no material anymore, you need to provide a shader+inputs+state

File: src/irr/asset/COBJMeshFileLoader.h [hunk 8]
```diff
@@ -170,24 +170,29 @@ class COBJMeshFileLoader : public asset::IAssetLoader
             SObjMtl() : Bumpiness (1.0f), Illumination(0),
                 RecalculateNormals(false)
             {
+				/*
+					Is it a dead code? If not, how can I set it?
```
Comment (devshgraphicsprogramming) [inline src/irr/asset/COBJMeshFileLoader.h:174]: useless and dead code.

File: src/irr/asset/COBJMeshFileLoader.h [hunk 9]
```diff
@@ -170,24 +170,29 @@ class COBJMeshFileLoader : public asset::IAssetLoader
             SObjMtl() : Bumpiness (1.0f), Illumination(0),
```
Comment (devshgraphicsprogramming) [inline src/irr/asset/COBJMeshFileLoader.h:170]: all of this material and texture stuff should move to a `CSpecializedShaderLoaderMTL`, `CPipelineLayoutLoaderMTL`, `CDescriptorSetLoaderMTL` (all read the same file, produce different assets that you'll need to put the mesh together)

File: src/irr/asset/CPLYMeshFileLoader.cpp [hunk 10]
```diff
@@ -67,7 +67,6 @@ asset::SAssetBundle CPLYMeshFileLoader::loadAsset(io::IReadFile* _file, const as
 
     SContext ctx;
 	ctx.File = std::move(core::make_smart_refctd_ptr<io::IReadFile>(_file));
```
Comment (devshgraphicsprogramming) [inline src/irr/asset/CPLYMeshFileLoader.cpp]: no need to refcount

File: include/irr/asset/ICPUMeshBuffer.h [hunk 11]
```diff
@@ -83,29 +83,36 @@ class ICPUMeshBuffer : public IMeshBuffer<ICPUBuffer, ICPUDescriptorSet, ICPURen
...
+	inline bool setVertexBufferBindingParams(uint32_t bindingIndex, uint32_t stride, E_VERTEX_INPUT_RATE inputRate = E_VERTEX_INPUT_RATE::EVIR_PER_VERTEX)
 	{
-		if (stride >= 2048ull || bindingIndex >= MAX_ATTR_BUF_BINDING_COUNT)
+		if (bindingIndex >= MAX_ATTR_BUF_BINDING_COUNT || stride >= 2048ull)
 			return false;
 
-		m_vertexBufferBindings[bindingIndex].buffer = std::move(bufferBinding.buffer);
-		m_vertexBufferBindings[bindingIndex].offset = bufferBinding.offset;
-
 		auto& binding(m_pipeline->getVertexInputParams().bindings[bindingIndex]);
```
Comment (devshgraphicsprogramming) [inline include/irr/asset/ICPUMeshBuffer.h]: don't initialize references through a constructor like syntax, use `=` always.

File: include/irr/asset/ICPUMeshBuffer.h [hunk 12]
```diff
@@ -83,29 +83,36 @@ class ICPUMeshBuffer : public IMeshBuffer<ICPUBuffer, ICPUDescriptorSet, ICPURen
...
 		binding.inputRate = inputRate;
 
 		return true;
 	}
+	inline bool setVertexBufferBinding(SBufferBinding&& bufferBinding, uint32_t bindingIndex)
+	{
+		if (bindingIndex >= MAX_ATTR_BUF_BINDING_COUNT)
+			return false;
+
+		m_vertexBufferBindings[bindingIndex].buffer = bufferBinding.buffer;
```
Comment (devshgraphicsprogramming) [inline include/irr/asset/ICPUMeshBuffer.h]: Do `m_vertexBufferBindings[bindingIndex] = std::move(bufferBinding);`

File: include/irr/asset/ICPUMeshBuffer.h [hunk 13]
```diff
@@ -83,29 +83,36 @@ class ICPUMeshBuffer : public IMeshBuffer<ICPUBuffer, ICPUDescriptorSet, ICPURen
...
+		m_vertexBufferBindings[bindingIndex].offset = bufferBinding.offset;
+
+		return true;
+	}
 	inline void setIndexBufferBinding(SBufferBinding&& bufferBinding)
 	{
-		m_indexBufferBinding.buffer = std::move(bufferBinding.buffer);
+		m_indexBufferBinding.buffer = bufferBinding.buffer;
```
Comment (devshgraphicsprogramming) [inline include/irr/asset/ICPUMeshBuffer.h]: `m_indexBufferBinding = std::move(bufferBinding);`

File: src/irr/asset/COBJMeshFileLoader.cpp [hunk 14]
```diff
@@ -561,14 +568,14 @@ const char* COBJMeshFileLoader::readTextures(const SContext& _ctx, const char* b
 	}
 	if (clamp)
     {
-        for (size_t i=0; i<_IRR_MATERIAL_MAX_TEXTURES_; i++)
```
Comment (devshgraphicsprogramming) [inline src/irr/asset/COBJMeshFileLoader.cpp]: you need to create sampler objects now instead.

File: src/irr/asset/COBJMeshFileLoader.cpp [hunk 15]
```diff
@@ -561,14 +568,14 @@ const char* COBJMeshFileLoader::readTextures(const SContext& _ctx, const char* b
...
-            currMaterial->Material.TextureLayer[i].SamplingParams.TextureWrapV = video::ETC_CLAMP_TO_EDGE;
-            currMaterial->Material.TextureLayer[i].SamplingParams.TextureWrapW = video::ETC_CLAMP_TO_EDGE;
+            currMaterial->Material.TextureLayer[i].SamplingParams.TextureWrapU = ISampler::E_TEXTURE_CLAMP::ETC_CLAMP_TO_EDGE;
+            currMaterial->Material.TextureLayer[i].SamplingParams.TextureWrapV = ISampler::E_TEXTURE_CLAMP::ETC_CLAMP_TO_EDGE;
+            currMaterial->Material.TextureLayer[i].SamplingParams.TextureWrapW = ISampler::E_TEXTURE_CLAMP::ETC_CLAMP_TO_EDGE;
         }
     }
-	for (size_t i = 0; i < _IRR_MATERIAL_MAX_TEXTURES_; i++)
+	for (size_t i = 0; i < video::OGL_STATE_MAX_TEXTURES; i++)
 		currMaterial->Material.TextureLayer[i].SamplingParams.AnisotropicFilter = 16u;
```
Comment (devshgraphicsprogramming) [inline src/irr/asset/COBJMeshFileLoader.cpp:583]: you need to create sampler objects now instead, in the CDescriptorSetLoaderMTL

And make sure you cache them in the `builtin` path (if caching flags allow)

**Pull request closed without merging**